src/storage/src/upsert/types.rs

```diff
@@ -623,6 +626,30 @@ where
     where
         G: IntoIterator<Item = UpsertKey>,
         R: IntoIterator<Item = &'r mut UpsertValueAndSize<O>>;
+
+    /// Merge values for all `merges` keys. The values for each key are merged into the existing
+    /// key value using the `snapshot_merge_function` when the backend decides to run the merge
+    /// operation. This allows avoiding the read-modify-write method of updating many values to
+    /// improve performance.
+    ///
+    /// TODO: Figure out stats for this
+    async fn multi_merge<P>(&mut self, merges: P) -> Result<PutStats, anyhow::Error>
+    where
+        P: IntoIterator<Item = (UpsertKey, StateValue<O>)>;
```

Comment:
does it make sense for this to take `StateValue`? Really it should take `(UpsertKey, UpsertValue, mz_repr::Diff)` like `merge_snapshot_chunk`, right?

Reply:
that would break the 'associativity' of the merge operator since the merge operands would be different types than the put/get value type. I could do that but it would require us to implement the separate full and partial merge operators.

I added the `val.merge_update` method that you said was 'absolutely fireeeeee' below to convert these values to a `StateValue` and allow us to keep the value-types the same for put and merge ops, but could be convinced to switch it if you think it would be a better method!

Reply:
actually it seems that I will still have to change this type to something that includes the `diff` to allow for the stats behavior you described below, so I'll be updating this slightly. However the type of the value we actually send in the `rocksdb.multi_update` call needs to be the same as the put value type if we want to avoid defining non-associative merge operators, so my comment above still stands

Reply:
> I added the val.merge_update method that you said was 'absolutely fireeeeee' below to convert these values to a StateValue and allow us to keep the value-types the same for put and merge ops, but could be convinced to switch it if you think it would be a better method!

Ah I see! Your plan sounds good!